packages/blockchain/src/index.ts

```diff
@@ -477,217 +510,90 @@ export default class Blockchain implements BlockchainInterface {
+    return await this.initAndLock<Buffer[]>(async () => {
+      let max: number
+      let mid: number
+      let min: number
+
+      max = hashes.length - 1
+      mid = min = 0
+
+      while (max >= min) {
+        let number
+        try {
+          number = await this.dbManager.hashToNumber(hashes[mid])
+        } catch (error) {
+          if (error.type !== 'NotFoundError') {
+            throw error
+          }
         }
-      })
-      // flag stale headBlock for reset
-      if (staleHash && this._headBlock?.equals(staleHash)) {
-        this._staleHeadBlock = true
-      }
-
-      try {
-        const parentHeader = await this._getHeader(header.parentHash, blockNumber.subn(1))
-        await this._rebuildCanonical(parentHeader, ops)
-      } catch (error) {
-        this._staleHeads = []
-        if (error.type !== 'NotFoundError') {
-          throw error
+        if (number) {
+          min = mid + 1
+        } else {
+          max = mid - 1
         }
+        mid = Math.floor((min + max) / 2)
       }
-    } else {
-      // set stale heads to last previously valid canonical block
-      this._staleHeads.forEach((name: string) => {
-        this._heads[name] = blockHash
-      })
-      this._staleHeads = []
-      // set stale headBlock to last previously valid canonical block
-      if (this._staleHeadBlock) {
-        this._headBlock = blockHash
-        this._staleHeadBlock = false
-      }
-    }
+      return hashes.slice(min)
+    })
   }
 
   /**
-   * Deletes a block from the blockchain. All child blocks in the chain are deleted and any
-   * encountered heads are set to the parent block.
-   *
+   * Completely deletes a block from the blockchain including any references to this block.
+   * If this block was in the canonical chain, then also each child block of this block is deleted
+   * Also, if this was a canonical block, each head header which is part of this now stale chain will be set to the parentHeader of this block
+   * An example reason to execute is when running the block in the VM invalidates this block: this will then reset the canonical head to the past block
+   * (which has been validated in the past by the VM, so we can be sure it is correct).
    * @param blockHash - The hash of the block to be deleted
    */
   async delBlock(blockHash: Buffer) {
-    if (!this._initDone) {
-      await this._init()
-    }
-
-    await this._lock.wait()
-    try {
-      await this._delBlock(blockHash)
-      this._lock.release()
-    } catch (error) {
-      this._lock.release()
-      throw error
-    }
+    // Q: is it safe to make this not wait for a lock?
+    // this is called from `runBlockchain` in case `runBlock` throws (i.e. the block is invalid). But is this the way to go?
+    // If we know this is called from the iterator/runBlockchain we are safe, but if this is called from anywhere else then this might lead to a concurrency problem?
```

Comment:
`delBlock` has no lock because VM calls it for deletion... do we want this? The alternative here is to delete the block and halt the iterator if the `onBlock` function in the iterator throws. We can thus internally call _delBlock and can put a lock on this one, which I think is safer. Then the only function without a lock is `getBlock`, which should be fine as the iterator blocks any `put` operation until it is done (so `BLOCKHASH` will read from current canonical chain)